sig-and-wg/sig-core/proposals/common-validator.md

```diff
@@ -0,0 +1,266 @@
+
+### Makefile Targets
+
+Makefile targets should be simple and should allow contributors to execute all required validations as one command. It should be unified across all `kyma-project` repositories regardless language or type.
+
+Currently, `makefile` in components looks like that:
+
+```makefile
+APP_NAME = ui-api-layer
+IMG = $(DOCKER_PUSH_REPOSITORY)$(DOCKER_PUSH_DIRECTORY)/$(APP_NAME)
+TAG = $(DOCKER_TAG)
+
+.PHONY: ci-pr ci-master ci-release resolve build-and-test build-image push-image
+
+ci-pr: resolve build-and-test build-image push-image
+ci-master: resolve build-and-test build-image push-image
+ci-release: resolve build-and-test build-image push-image
+
+resolve:
+	dep ensure -v -vendor-only
+build-and-test:
+	./before-commit.sh ci
+build-image:
+	docker build -t $(APP_NAME) .
+push-image:
+	docker tag $(APP_NAME) $(IMG):$(TAG)
+	docker push $(IMG):$(TAG)
+```
+
+Only the `ci-*` targets are unified in `kyma-project` organization, because they are required by CI system. Other targets may have different names, because they are not unified and not required by any system. For example `ui-api-layer` has `build-and-test` target, `binding-usage-controller` has `build` - both targets executes `before-commit.sh` script. In such case contributor doesn't know how to work with components in one repository.
+
+As a solution for that I would like to unify targets:
+ - `ci-pr`, `ci-master`, `ci-release` - targets required by CI system
+ - `validate` - execute all validation
+ - `resolve` - download dependencies
+ - `build` - execute build of component
+ - `test` - execute tests for component
+ - `lint` - execute linters for component
+ - `format` - execute code formating for component
+ - `build-image` - build docker image with component
+ - `push-image` - push docker image to repository
+
+ After unification, `makefile` for `ui-api-layer` component will look like that:
+
+ ```makefile
+APP_NAME = ui-api-layer
+IMG = $(DOCKER_PUSH_REPOSITORY)$(DOCKER_PUSH_DIRECTORY)/$(APP_NAME)
+TAG = $(DOCKER_TAG)
+
+.PHONY: ci-pr ci-master ci-release resolve build test format lint build-image push-image
+
+ci-pr: validate build-image push-image
+ci-master: ci-pr
+ci-release: ci-master
+
+validate: resolve build test format lint
+
+resolve:
+	dep ensure -v -vendor-only
+build:
+	go build -o $(APP_NAME)
+test:
+	go test ./...
+format:
+	go fmt
+lint:
+	go vet
+
+build-image:
+	docker build -t $(APP_NAME) .
+push-image:
+	docker tag $(APP_NAME) $(IMG):$(TAG)
+	docker push $(IMG):$(TAG)
+ ```
+
+### Common Makefile
+
+As in previous point unified targets were introduced it would be fine to not duplicate all that `makefile` content in other components. Fortunately it is possible, because it is possible to include `makefiles`. Thanks to that we can create one common `makefile` that will be included in components `makefiles`.
+
+Lets name this common `makefile` `template.mk` - files with `*.mk` are threated as `makefiles` - and put it in `kyma/common` directory.
```

Comment:
I'm not a fan of placing it there. `kyma/comon` is a regular go library and has very different purpose then storing templates. I'd suggest placing it at root level.

Reply:
Good point, `kyma/common` was just an example location. I think it shouldn't be in the `root` of the repository, because we may want to introduce also some common scripts for validating or something in the future.

Reply:
I would also rename the file to `template.go.mk` as we have other technologies also here, like bash scripts etc. They are not something we are proud of, but we cannot just ignore our children ;)